presto-main/src/main/java/com/facebook/presto/operator/BlockEncodingBuffers.java

```diff
@@ -0,0 +1,363 @@
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.DictionaryBlock;
+import com.facebook.presto.spi.block.LongArrayBlock;
+import com.facebook.presto.spi.block.RunLengthEncodedBlock;
+import io.airlift.slice.SliceOutput;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+
+import static com.facebook.presto.array.Arrays.ensureCapacity;
+import static com.facebook.presto.operator.UncheckedByteArrays.setByteUnchecked;
+import static com.facebook.presto.spi.MoreByteArrays.fill;
+import static com.google.common.base.Verify.verify;
+import static io.airlift.slice.SizeOf.SIZE_OF_BYTE;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public abstract class BlockEncodingBuffers
+{
+    private static final int BITS_IN_BYTE = 8;
+
+    // The block after peeling off the Dictionary or RLE wrappings.
+    protected Block decodedBlock;
+
+    // The number of positions (rows) to be copied from the incoming block.
+    protected int positionCount;
+
+    // The number of positions (rows) to be copied from the incoming block within current batch.
+    protected int batchSize;
+
+    // The offset into positions/mappedPositions array that the current batch starts copying.
+    // I.e. in each batch we copy the values of rows from positions[positionsOffset] to positions[positionsOffset + batchSize]
+    protected int positionsOffset;
+
+    // The number of positions (rows) buffered so far
+    protected int bufferedPositionCount;
+
+    // Whether the positions array was mapped to mappedPositions
+    protected boolean positionsMapped;
+
+    // The positions (row numbers) of the incoming Block to be copied to this buffer.
+    // All top level BlockEncodingBuffers for the same partition share the same positions array.
+    private int[] positions;
+
+    // The mapped positions of the incoming Dictionary or RLE Block
+    @Nullable
+    private int[] mappedPositions;
+
+    @Nullable
+    private byte[] nullsBuffer;
+
+    // The address offset in the nullsBuffer if new values are to be added.
+    private int nullsBufferIndex;
+
+    // For each batch we put the nulls values up to a multiple of 8 into nullsBuffer. The rest is kept in remainingNulls.
+    private final boolean[] remainingNulls = new boolean[BITS_IN_BYTE];
+
+    // Number of null values not put into nullsBuffer from last batch
+    private int remainingNullsCount;
+
+    // Boolean indicating whether there are any null values in the nullsBuffer. It is possible that all values are non-null.
+    private boolean hasEncodedNulls;
+
+    public static BlockEncodingBuffers createBlockEncodingBuffers(DecodedBlockNode decodedBlockNode)
+    {
+        requireNonNull(decodedBlockNode, "decodedBlockNode is null");
+
+        Object decodedBlock = decodedBlockNode.getDecodedBlock();
```

Comment:
nit: Can the type be `Block` instead of `Object`?

Reply:
decodedBlock could be a block or ColumnarXXX so we can't use Block for all types. For primitive types we can make DecodedBlockNode only contain Block only and thus can use Block type here, but managing the changes among commits takes too much time therefore I just used Object here. If you have strong preference using Block for primitive types I'll make the change.

Reply:
@yingsu00 : This makes sense. Just adding a comment about why use `Object` -- since other engineers will likely have the same question when first read the code :)

Reply:
> @yingsu00 : This makes sense. Just adding a comment about why use `Object` -- since other engineers will likely have the same question when first read the code :)

Sure. I added the following comment:
```
        // decodedBlock could be a block or ColumnarArray/Map/Row
        Object decodedBlock = decodedBlockNode.getDecodedBlock();
```